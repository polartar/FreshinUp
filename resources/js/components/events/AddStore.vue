<template>
  <div>
    <div class="px-4 pt-4">
      <v-text-field
        v-model="searchText"
        label="Search by Fleet Member name"
        single-line
        outline
      />
      <div>
        <div class="pb-4">
          <button
            style="font-weight: bold; color: lightslategray; outline: 0;"
            @click="toggleShowFilter"
          >
            <span v-if="showFilters">
              <v-flex>
                <v-icon small>fa-caret-down</v-icon>
                Hide Filters
              </v-flex>
            </span>
            <span v-else>
              <v-flex>
                <v-icon small>fa-caret-right</v-icon>
                Show Filters
              </v-flex>
            </span>
          </button>
        </div>
        <div
          v-show="showFilters"
        >
          <v-layout
            row
            wrap
          >
            <v-flex
              sm3
              pr-2
            >
              <v-select
                v-model="selectedState"
                :items="locations"
                label="State of incorporation"
                outline
                single-line
              />
            </v-flex>
            <v-flex
              sm3
              pr-2
            >
              <v-select
                v-model="selectedType"
                :items="storeTypes"
                item-value="id"
                item-text="name"
                label="Type"
                outline
                single-line
              />
            </v-flex>
            <v-flex
              sm3
              pr-2
            >
              <v-select
                v-model="selectedTags"
                :items="tags"
                label="Tags"
                outline
                single-line
                multiple
              />
            </v-flex>
            <v-flex
              sm3
            >
              <v-btn
                large
                depressed
                @click="clearAllFilters"
              >
                Clear all filters
              </v-btn>
            </v-flex>
          </v-layout>
        </div>
      </div>
    </div>
    <div
      class="px-4"
      style="border-top: 1px solid gainsboro;"
    >
      <v-data-table
        v-model="selected"
        :headers="headers"
        :items="filteredStores"
        :search="searchText"
        item-key="uuid"
        hide-actions
        select-all
        :pagination.sync="pagination"
      >
        <template v-slot:items="props">
          <td>
            <v-checkbox
              v-model="props.selected"
              primary
              hide-details
            />
          </td>
          <td
            class="py-2"
            nowrap
          >
            <div style="position: relative;">
              <v-tooltip top>
                <template v-slot:activator="{ on, attrs }">
                  <router-link
                    class="primary--text"
                    :to="{ path: '/admin/fleet-members/'}"
                    target="_blank"
                    v-bind="attrs"
                    v-on="on"
                  >
                    {{ get(props, 'item.name') }}
                  </router-link>
                </template>
                View fleet member details (new tab)
              </v-tooltip>
            </div>
            <div
              class="grey--text text--darken-2"
            >
              {{ get(storeTypesById, 'props.item.type_id') }}
            </div>
          </td>
          <td class="py-2">
            {{ get(props, 'item.state_of_incorporation') }}
          </td>
          <td class="py-2">
            <div>
              <v-chip
                v-for="(tag, index) of get(props, 'item.tags', [])"
                :key="index"
                color="secondary"
                text-color="white"
              >
                {{ tag }}
              </v-chip>
            </div>
          </td>
          <td class="py-2">
            <v-btn
              :depressed="manageButtonLabel(props.item) !== 'Assign'"
              :disabled="!isEligible(props.item)"
              :class="manageButtonClass(props.item)"
              @click="onManageClicked(props.item)"
            >
              {{ manageButtonLabel(props.item) }}
            </v-btn>
          </td>
        </template>
      </v-data-table>
      <div class="text-xs-center pt-2">
        <v-pagination
          v-model="pagination.page"
          :length="pages"
        />
      </div>
    </div>
  </div>
</template>
<script>
import get from 'lodash/get'
import _uniq from 'lodash/uniq'

export const HEADERS = [
  { text: 'Fleet member', value: 'name' },
  { text: 'State of incorporation', value: 'state_of_incorporation' },
  { text: 'Tags', value: 'tags' },
  { text: 'Manage' }
]
export default {
  props: {
    stores: {
      type: Array,
      default: () => []
    },
    event: {
      type: Object,
      required: true
    },
    storeTypes: {
      type: Array,
      default: () => []
    }
  },
  data () {
    return {
      searchText: '',
      selectedState: '',
      selectedType: '',
      selectedTags: [],
      pagination: {
        rowsPerPage: 5,
        page: 1
      },
      page: 1,
      showFilters: false,
      selected: [],
      headers: HEADERS
    }
  },
  computed: {
    pages () {
      if (this.pagination.rowsPerPage == null ||
          this.totalItems == null
      ) return 0

      return Math.ceil(this.totalItems / this.pagination.rowsPerPage)
    },

    filteredStores () {
      let items = this.stores

      if (this.selectedState) {
        items = items.filter(i => i['state_of_incorporation'] === this.selectedState)
      }

      if (this.selectedType) {
        items = items.filter(i => i['type_id'] === this.selectedType)
      }

      if (this.selectedTags.length) {
        items = items.filter(i => i['tags'].some(tag => this.selectedTags.includes(tag)))
      }

      return items
    },
    totalItems () {
      return this.filteredStores.length
    },
    locations () {
      return _uniq(this.stores.map(m => m['state_of_incorporation']))
    },
    tags () {
      const ts = []
      this.stores.forEach(m => ts.push(...m['tags']))
      return _uniq(ts)
    },
    storeTypesById () {
      return this.storeTypes.reduce((map, type) => {
        map[type.id] = type
        return map
      }, {})
    }
  },
  methods: {
    get,
    toggleShowFilter () {
      this.showFilters = !this.showFilters
    },

    clearAllFilters () {
      this.selectedState = ''
      this.selectedType = ''
      this.selectedTags = []
    },
    hasBookedAnEvent (member) {
      return member.events.findIndex(e => e.uuid !== this.event.uuid) !== -1
    },
    isEligible (member) {
      return !this.hasBookedAnEvent(member) && !member['has_expired_licences_docs']
    },
    isAssignedToThisEvent (member) {
      return member.events.findIndex(e => e.uuid === this.event.uuid) !== -1
    },
    isDeclined (member) {
      return member.events.findIndex(e => e.uuid === this.event.uuid && e.declined) !== -1
    },
    manageButtonLabel (item) {
      if (this.isEligible(item)) {
        if (this.isDeclined(item)) { return 'Declined' }
        if (!this.isAssignedToThisEvent(item)) { return 'Assign' } else { return 'Assigned' }
      }

      if (this.hasBookedAnEvent(item)) { return 'Booked' }

      if (item['has_expired_licences_docs']) { return 'Expired' }

      return ''
    },
    onManageClicked (item) {
      if (this.manageButtonLabel(item) === 'Assign') {
        this.$emit('assign', item)
      }
    },
    manageButtonClass (item) {
      const label = this.manageButtonLabel(item)
      return {
        'primary': label === 'Assign',
        'blue-grey lighten-5': ['Expired', 'Booked'].includes(label),
        'blue-grey lighten-3 white--text': ['Declined', 'Assigned'].includes(label)
      }
    }
  }
}
</script>
