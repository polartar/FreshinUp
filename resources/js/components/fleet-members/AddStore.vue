<template>
  <div>
    <div class="px-2 py-2">
      <filter-sorter
        :types="types"
        @input="runFilters"
      />
    </div>
    <div
      class="px-2"
      style="border-top: 1px solid gainsboro;"
    >
      <f-data-table
        :headers="headers"
        :items="stores"
        :multi-item-actions="multipleItemActions"
        item-key="uuid"
        hide-actions
        select-all
        v-bind="$attrs"
        v-on="$listeners"
      >
        <template v-slot:item-inner-name="{ item }">
          <div style="position: relative;">
            <v-tooltip top>
              <template v-slot:activator="{ on, attrs }">
                <a
                  class="primary--text"
                  href="#"
                  @click.prevent="viewItem"
                >
                  {{ get(item, 'name') }}
                </a>
              </template>
              View fleet member details (new tab)
            </v-tooltip>
          </div>
        </template>
        <template v-slot:item-inner-tags="{ item }">
          <div>
            <v-chip
              v-for="(tag, index) of get(item, 'tags', [])"
              :key="index"
              color="secondary"
              text-color="white"
            >
              {{ tag.name }}
            </v-chip>
          </div>
        </template>
        <template v-slot:item-inner-state_of_incorporation="{ item }">
          <div class="grey--text">
            {{ get(item, 'state_of_incorporation') }}
          </div>
        </template>
        <template v-slot:item-inner-manage="{ item }">
          <v-btn
            :depressed="manageButtonLabel(item) !== 'Assign'"
            :disabled="!isEligible(item)"
            :class="manageButtonClass(item)"
            @click="onManageClicked(item)"
          >
            {{ manageButtonLabel(item) }}
          </v-btn>
        </template>
      </f-data-table>
    </div>
  </div>
</template>
<script>
import get from 'lodash/get'
import FDataTable from '@freshinup/core-ui/src/components/FDataTable'
import AddStoreFilterSorter from './AddStoreFilterSorter'

export const HEADERS = [
  { text: 'Fleet member', value: 'name' },
  { text: 'State of incorporation', value: 'state_of_incorporation' },
  { text: 'Tags', value: 'tags' },
  { text: 'Manage', value: 'manage' }
]

export const MULTIPLE_ITEM_ACTIONS = []
export default {
  components: {
    FDataTable,
    FilterSorter: AddStoreFilterSorter
  },
  props: {
    headers: { type: Array, default: () => HEADERS },
    multipleItemActions: { type: Array, default: () => MULTIPLE_ITEM_ACTIONS },
    event: { type: Object, required: true, default: () => ({}) },
    stores: { type: Array, default: () => [] },
    types: { type: Array, default: () => [] }
  },
  data () {
    return {
      showFilters: false
    }
  },
  methods: {
    get,
    runFilters (payload) {
      this.$emit('run-filter', payload)
    },
    viewItem (item) {
      this.$emit('manage', 'view', item)
      this.$emit('manage-view', item)
    },
    // TODO: methods need more attention at some point later in the future
    hasBookedAnEvent (member) {
      const stores = get(member, 'event_stores', [])
      return stores.findIndex(e => e.event_uuid !== this.event.uuid) !== -1
    },
    isEligible (member) {
      return !this.hasBookedAnEvent(member) && !member['has_expired_licences_docs']
    },
    isAssignedToThisEvent (member) {
      const stores = get(member, 'event_stores', [])
      return stores.findIndex(e => e.event_uuid === this.event.uuid) !== -1
    },
    isDeclined (member) {
      const stores = get(member, 'event_stores', [])
      return stores.findIndex(e => e.event_uuid === this.event.uuid && e.declined) !== -1
    },
    manageButtonLabel (item) {
      if (this.isEligible(item)) {
        if (this.isDeclined(item)) { return 'Declined' }
        if (!this.isAssignedToThisEvent(item)) { return 'Assign' } else { return 'Assigned' }
      }
      if (this.hasBookedAnEvent(item)) { return 'Booked' }
      if (item['has_expired_licences_docs']) { return 'Expired' }
      return ''
    },
    onManageClicked (item) {
      if (this.manageButtonLabel(item) === 'Assign') {
        this.$emit('manage', 'assign', item)
        this.$emit('manage-assign', item)
      }
    },
    manageButtonClass (item) {
      const label = this.manageButtonLabel(item)
      return {
        'primary': label === 'Assign',
        'blue-grey lighten-5': ['Expired', 'Booked'].includes(label),
        'blue-grey lighten-3 white--text': ['Declined', 'Assigned'].includes(label)
      }
    }
  }
}
</script>
