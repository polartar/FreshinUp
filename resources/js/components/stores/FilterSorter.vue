<template>
  <search-filter-sorter
    ref="filter"
    class="filter-transparent"
    expanded
    without-filter-label
    sort-label="Sort by"
    :sort-options="sortables"
    color="transparent"
    v-bind="$attrs"
    v-on="$listeners"
    @run="run"
    @clear="clearFilters"
  >
    <template v-slot:expanded="slotProps">
      <v-card-text class="px-0 pb-0">
        <v-layout
          row
          justify-space-between
        >
          <v-flex>
            <v-layout
              row
              justify-space-between
              align-center
              mb-2
            >
              <filter-label>
                Statuses
              </filter-label>
              <clear-button
                v-if="filters.status && filters.status.length > 0"
                color="white"
                @clear="filters.status = null;"
              />
            </v-layout>
            <multi-select
              v-model="filters.status"
              placeholder="Select Status"
              :items="statuses"
              item-value="id"
              item-text="name"
              select-all-name="All Status"
              solo
              flat
              hide-details
            />
          </v-flex>
          <v-flex
            ml-2
          >
            <v-layout
              row
              justify-space-between
              mb-2
            >
              <filter-label>
                Supplier
              </filter-label>
              <clear-button
                v-if="filters.supplier_uuid && filters.supplier_uuid.length > 0"
                color="white"
                @clear="filters.supplier_uuid = null; $refs.host.resetTerm()"
              />
            </v-layout>
            <multi-simple
              ref="supplier"
              v-model="filters.supplier_uuid"
              url="companies?filter[type_key]=supplier"
              term-param="filter[name]"
              results-id-key="uuid"
              placeholder="Select"
              background-color="white"
              class="mt-0 pt-0"
              height="48"
              not-clearable
              solo
              flat
            />
          </v-flex>
          <v-flex
            ml-2
          >
            <v-layout
              row
              justify-space-between
              mb-2
            >
              <filter-label>
                Hometown
              </filter-label>
              <clear-button
                v-if="filters.location_uuid"
                color="white"
                @clear="filters.location_uuid = null; $refs.location.resetTerm()"
              />
            </v-layout>
            <simple
              ref="location"
              url="foodfleet/locations"
              term-param="filter[name]"
              results-id-key="uuid"
              :value="filters.location_uuid"
              placeholder="Select"
              background-color="white"
              class="mt-0 pt-0"
              height="48"
              not-clearable
              solo
              flat
              @input="selectLocation"
            />
          </v-flex>
          <v-flex
            ml-2
          >
            <v-layout
              row
              justify-space-between
              mb-2
            >
              <filter-label>
                Tags
              </filter-label>
              <clear-button
                v-if="filters.store_tag_uuid && filters.store_tag_uuid.length > 0"
                color="white"
                @clear="filters.store_tag_uuid = null; $refs.tag.resetTerm()"
              />
            </v-layout>
            <multi-simple
              ref="tag"
              v-model="filters.store_tag_uuid"
              url="foodfleet/store-tags"
              term-param="filter[name]"
              results-id-key="uuid"
              placeholder="Search Tag"
              background-color="white"
              class="mt-0 pt-0"
              height="48"
              not-clearable
              solo
              flat
            />
          </v-flex>
        </v-layout>
      </v-card-text>
    </template>
  </search-filter-sorter>
</template>
<script>
import ClearButton from '~/components/ClearButton'
import FilterLabel from '~/components/FilterLabel'
import SearchFilterSorter from 'fresh-bus/components/search/filter-sorter'
import Simple from 'fresh-bus/components/search/simple'
import MultiSelect from 'fresh-bus/components/ui/FMultiSelect'
import MultiSimple from 'fresh-bus/components/ui/FMultiSimple'
export default {
  components: {
    Simple,
    MultiSimple,
    FilterLabel,
    ClearButton,
    MultiSelect,
    SearchFilterSorter
  },
  props: {
    filters: {
      type: Object,
      default: () => ({
        status: null,
        store_tag_uuid: null,
        location_uuid: null,
        supplier_uuid: null
      })
    },
    statuses: {
      type: Array,
      default: () => []
    },
    sortables: {
      type: Array,
      default: () => []
    }
  },
  watch: {
    filters: {
      handler: function (value) {
        const params = this.$refs.filter.getRunParams()
        this.run(params)
      },
      deep: true
    }
  },
  methods: {
    run (params) {
      let finalParams = {
        name: params.term,
        ...this.filters
      }
      if (this.filters.store_tag_uuid) {
        finalParams.store_tag_uuid = this.filters.store_tag_uuid.map(item => item.uuid)
      }
      if (this.filters.supplier_uuid) {
        finalParams.supplier_uuid = this.filters.supplier_uuid.map(item => item.uuid)
      }
      this.$emit('runFilter', finalParams)
    },
    selectLocation (location) {
      this.filters.location_uuid = location ? location.uuid : null
    },
    clearFilters (params) {
      this.$refs.tag.resetTerm()
      this.$refs.supplier.resetTerm()
      this.$refs.location.resetTerm()
      this.filters.status = this.filters.store_tag_uuid = this.filters.location_uuid = this.filters.supplier_uuid = null
    }
  }
}
</script>
<style lang="styl" scoped>
  /deep/ .data-time-picker.no-border input.field-input{
    border: none !important;
  }
  /deep/ .filter-sorter-expanded-layout{
    align-items: flex-end;
  }
  /deep/ .filter-sorter-expanded-layout>.flex.text-no-wrap>.v-btn{
    margin: 0;
    height: 48px;
  }
  .filter-transparent{
    box-shadow: none;
  }
</style>
