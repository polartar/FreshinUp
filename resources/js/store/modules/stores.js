import isObject from 'lodash/isObject'
import omitBy from 'lodash/omitBy'
import makeRestStore, {
  buildApi,
  makeModule
} from '@freshinup/core-ui/src/store/utils/makeRestStore'
import pick from 'lodash/pick'

export default ({ items, item }) => {
  const storesApi = buildApi('foodfleet/stores', { items, item })
  const store = makeModule(storesApi.getStore(), 'stores')

  const sortables = [
    { value: '-created_at', text: 'Newest' },
    { value: 'created_at', text: 'Oldest' },
    { value: 'name', text: 'Name (A - Z)' },
    { value: '-name', text: 'Name (Z - A)' }
  ]

  const FILLABLES = [
    'id',
    'uuid',
    'name',
    'square_id',
    'status_id',
    'type_id',
    'supplier_uuid',
    'address_uuid',
    'website',
    'contact_phone',
    'size',
    'image',
    'created_at',
    'updated_at',
    'deleted_at',
    'owner_uuid',
    'pos_system',
    'size_of_truck_trailer',
    'phone',
    'state_of_incorporation',
    'facebook',
    'twitter',
    'instagram',
    'staff_notes',
    'tags'
  ]

  // Initial State
  store.state = {
    ...store.state,
    sortables
  }

  const _createItem = store.actions.createItem
  const _updateItem = store.actions.updateItem
  const process = (payload) => {
    const data = pick(payload.data, FILLABLES)
    if (data.tags) {
      data.tags = data.tags.map(tag => {
        return isObject(tag) ? tag.uuid : tag
      })
    }
    return Object.assign({}, payload, { data })
  }
  store.actions = {
    ...store.actions,
    createItem (context, payload) {
      const processedPayload = process(payload)
      processedPayload.data = omitBy(processedPayload.data, (value) => {
        return String(value).length === 0
      })
      return _createItem(context, processedPayload)
    },
    updateItem (context, payload) {
      const processedPayload = process(payload)
      return _updateItem(context, processedPayload)
    }
  }

  return {
    namespaced: true,
    ...store,
    modules: {
      summary: makeRestStore(
        'summary',
        { item },
        {
          itemPath: ({ id }) => `/foodfleet/store-summary/${id}`
        }
      ),
      serviceSummary: makeRestStore(
        'serviceSummary',
        { item },
        {
          itemPath: ({ id }) => `/foodfleet/store-service-summary/${id}`
        }
      )
    }
  }
}
