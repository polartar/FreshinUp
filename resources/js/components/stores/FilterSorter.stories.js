import { storiesOf } from '@storybook/vue'
import { action } from '@storybook/addon-actions'
import MockAdapter from 'axios-mock-adapter'
import axios from 'axios/index'

// Components
import FilterSorter from './FilterSorter.vue'

let statuses = [
  { value: 1, text: 'Draft' },
  { value: 2, text: 'Pending' },
  { value: 3, text: 'Revision' },
  { value: 4, text: 'Rejected' },
  { value: 5, text: 'Approved' },
  { value: 6, text: 'On hold' }
]

let sortables = [
  { value: '-created_at', text: 'Newest' },
  { value: 'created_at', text: 'Oldest' },
  { value: 'title', text: 'Title (A - Z)' },
  { value: '-title', text: 'Title (Z - A)' }
]

const mock = new MockAdapter(axios)
mock.onGet('foodfleet/store-tags').reply(200, {
  data: [
    { uuid: 1, name: 'Store Tag 1' },
    { uuid: 2, name: 'Store Tag 2' },
    { uuid: 3, name: 'Store Tag 3' },
    { uuid: 4, name: 'Store Tag 4' }
  ]
})
mock.onGet('foodfleet/locations').reply(200, {
  data: [
    { uuid: 1, name: 'Location 1' },
    { uuid: 2, name: 'Location 2' },
    { uuid: 3, name: 'Location 3' },
    { uuid: 4, name: 'Location 4' }
  ]
})

storiesOf('FoodFleet|fleet-member/FilterSorter', module)
  .addParameters({
    backgrounds: [
      { name: 'default', value: '#f1f3f6', default: true }
    ]
  })
  .add('default', () => ({
    components: { FilterSorter },
    methods: {
      filterStores (params) {
        action('Run')(params)
      }
    },
    template: `
      <v-container>
        <filter-sorter
          @runFilter="filterStores"
        />
      </v-container>
    `
  }))
  .add('with statuses', () => ({
    components: { FilterSorter },
    data () {
      return {
        statuses: statuses
      }
    },
    methods: {
      filterStores (params) {
        action('Run')(params)
      }
    },
    template: `
      <v-container>
        <filter-sorter
          :statuses="statuses"
          @runFilter="filterStores"
        />
      </v-container>
    `
  }))
  .add('with sortables', () => ({
    components: { FilterSorter },
    data () {
      return {
        sortables: sortables
      }
    },
    methods: {
      filterStores (params) {
        action('Run')(params)
      }
    },
    template: `
      <v-container>
        <filter-sorter
          :sortables="sortables"
          @runFilter="filterStores"
        />
      </v-container>
    `
  }))
