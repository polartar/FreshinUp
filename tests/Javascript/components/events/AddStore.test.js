import { createLocalVue, mount, shallowMount } from '@vue/test-utils'
import Component from '~/components/events/AddStore.vue'

import * as Stories from '~/components/events/AddStore.stories'
import { FIXTURE_STORES } from '../../__data__/stores'
import { FIXTURE_STORE_TYPES } from '../../__data__/storeTypes'

describe('Add member (store) in event component', () => {
  let localVue
  describe('Snapshots', () => {
    test('Default', async () => {
      const wrapper = mount(Stories.Default())
      await wrapper.vm.$nextTick()
      expect(wrapper.element).toMatchSnapshot()
    })
    test('WithData', async () => {
      const wrapper = mount(Stories.WithData())
      await wrapper.vm.$nextTick()
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('Computed', () => {
    beforeEach(() => {
      localVue = createLocalVue()
    })

    test('pages', () => {
      const wrapper = shallowMount(Component, {
        localVue: localVue
      })
      wrapper.setProps({ stores: FIXTURE_STORES })
      wrapper.setData({
        pagination: {
          rowsPerPage: 5,
          page: 1
        }
      })
      expect(wrapper.vm.pages).toBe(1)

      wrapper.setData({
        pagination: {
          rowsPerPage: 3,
          page: 1
        }
      })
      expect(wrapper.vm.pages).toBe(2)
    })

    test('filteredStores', () => {
      const wrapper = shallowMount(Component, {
        localVue: localVue
      })

      wrapper.setProps({ stores: FIXTURE_STORES })
      wrapper.setData({
        selectedState: '',
        selectedType: '',
        selectedTags: []
      })

      expect(wrapper.vm.filteredStores).toHaveLength(5)
      expect(wrapper.vm.filteredStores).toEqual(FIXTURE_STORES)

      wrapper.setData({
        selectedState: 'New York',
        selectedType: '',
        selectedTags: []
      })

      expect(wrapper.vm.filteredStores).toHaveLength(1)
      expect(wrapper.vm.filteredStores).toEqual([FIXTURE_STORES[1]])

      wrapper.setData({
        selectedState: '',
        selectedType: 2,
        selectedTags: []
      })

      expect(wrapper.vm.filteredStores).toHaveLength(4)
      expect(wrapper.vm.filteredStores).toEqual(FIXTURE_STORES.filter(s => s.type_id === 2))

      wrapper.setData({
        selectedState: '',
        selectedType: '',
        selectedTags: [FIXTURE_STORES[0].tags[0]]
      })

      expect(wrapper.vm.filteredStores).toHaveLength(3)
    })

    test('totalItems', () => {
      const wrapper = shallowMount(Component, {
        localVue: localVue
      })

      wrapper.setProps({ stores: FIXTURE_STORES })

      expect(wrapper.vm.totalItems).toBe(5)
    })

    test('locations', () => {
      const wrapper = shallowMount(Component, {
        localVue: localVue
      })

      wrapper.setProps({ stores: FIXTURE_STORES })

      const expectedLocations = ['California', 'New York']

      expect(wrapper.vm.locations).toEqual(expectedLocations)
    })

    test('tags', () => {
      const wrapper = shallowMount(Component, {
        localVue: localVue
      })

      wrapper.setProps({ stores: FIXTURE_STORES })
      expect(wrapper.vm.tags).toHaveLength(2)
    })

    test('storeTypesById', async () => {
      const wrapper = shallowMount(Component, {
        localVue: localVue
      })

      wrapper.setProps({
        storeTypes: FIXTURE_STORE_TYPES
      })
      await wrapper.vm.$nextTick()
      const expected = FIXTURE_STORE_TYPES.reduce((map, type) => {
        map[type.id] = type
        return map
      }, {})
      expect(wrapper.vm.storeTypesById).toMatchObject(expected)
    })
  })

  describe('Methods', () => {
    beforeEach(() => {
      localVue = createLocalVue()
    })

    test('showFilters()', () => {
      const wrapper = shallowMount(Component, {
        localVue: localVue
      })

      wrapper.setData({
        showFilters: false
      })

      expect(wrapper.vm.showFilters).toBeFalsy()

      wrapper.vm.toggleShowFilter()
      expect(wrapper.vm.showFilters).toBeTruthy()

      wrapper.vm.toggleShowFilter()
      expect(wrapper.vm.showFilters).toBeFalsy()
    })

    test('clearAllFilters()', () => {
      const wrapper = shallowMount(Component, {
        localVue: localVue
      })

      wrapper.setData({
        selectedState: 'aaa',
        selectedType: 'ttt',
        selectedTags: ['a', 'b']
      })

      wrapper.vm.clearAllFilters()
      expect(wrapper.vm.selectedState).toBe('')
      expect(wrapper.vm.selectedType).toBe('')
      expect(wrapper.vm.selectedTags).toEqual([])
    })

    test('hasBookedAnEvent(store)', () => {
      const wrapper = shallowMount(Component, {
        localVue: localVue,
        propsData: {
          event: FIXTURE_STORES[0].events[0]
        }
      })

      expect(wrapper.vm.hasBookedAnEvent(FIXTURE_STORES[3])).toBeTruthy()
      expect(wrapper.vm.hasBookedAnEvent(FIXTURE_STORES[2])).toBeFalsy()
    })

    test('isEligible(store)', () => {
      const wrapper = shallowMount(Component, {
        localVue: localVue
      })
      wrapper.setProps({
        event: FIXTURE_STORES[0].events[0]
      })

      expect(wrapper.vm.isEligible(FIXTURE_STORES[0])).toBeTruthy()
      expect(wrapper.vm.isEligible(FIXTURE_STORES[1])).toBeTruthy()
      expect(wrapper.vm.isEligible(FIXTURE_STORES[2])).toBeFalsy()
    })

    test('isAssignedToThisEvent(store)', () => {
      const wrapper = shallowMount(Component, {
        localVue: localVue
      })

      wrapper.setProps({
        event: FIXTURE_STORES[0].events[0]
      })

      expect(wrapper.vm.isAssignedToThisEvent(FIXTURE_STORES[0])).toBeTruthy()
      expect(wrapper.vm.isAssignedToThisEvent(FIXTURE_STORES[4])).toBeTruthy()
      expect(wrapper.vm.isAssignedToThisEvent(FIXTURE_STORES[1])).toBeFalsy()
    })

    test('isDeclined(store)', () => {
      const wrapper = shallowMount(Component, {
        localVue: localVue
      })

      wrapper.setProps({
        event: FIXTURE_STORES[4].events[0]
      })

      expect(wrapper.vm.isDeclined(FIXTURE_STORES[4])).toBeTruthy()
      expect(wrapper.vm.isDeclined(FIXTURE_STORES[1])).toBeFalsy()
    })

    test('manageButtonLabel(store)', () => {
      const wrapper = shallowMount(Component, {
        localVue: localVue
      })

      wrapper.setProps({
        event: FIXTURE_STORES[0].events[0]
      })

      expect(wrapper.vm.manageButtonLabel(FIXTURE_STORES[0])).toBe('Assigned')
      expect(wrapper.vm.manageButtonLabel(FIXTURE_STORES[1])).toBe('Assign')
      expect(wrapper.vm.manageButtonLabel(FIXTURE_STORES[2])).toBe('Expired')
      expect(wrapper.vm.manageButtonLabel(FIXTURE_STORES[3])).toBe('Booked')
      expect(wrapper.vm.manageButtonLabel(FIXTURE_STORES[4])).toBe('Declined')
    })

    test('manageButtonClass(store)', () => {
      const wrapper = shallowMount(Component, {
        localVue: localVue
      })

      wrapper.setProps({
        event: FIXTURE_STORES[4].events[0]
      })

      const expectedClassObject = {
        'blue-grey lighten-3 white--text': true,
        'primary': false,
        'blue-grey lighten-5': false
      }

      expect(wrapper.vm.manageButtonClass(FIXTURE_STORES[4])).toMatchObject(expectedClassObject)
    })
  })
})
