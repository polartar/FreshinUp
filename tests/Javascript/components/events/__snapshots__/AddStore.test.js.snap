// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add member (store) in event component Snapshot defaults 1`] = `
<div>
  <div
    class="px-4 pt-4"
  >
    <v-text-field-stub
      label="Search by Fleet Member name"
      outline=""
      single-line=""
      value=""
    />
     
    <div>
      <div
        class="pb-4"
      >
        <button
          style="font-weight: bold; color: lightslategray; outline: 0;"
        >
          
          Show Filters
        
        </button>
      </div>
       
      <div
        style="display: none;"
      >
        <v-layout-stub
          row=""
          tag="div"
          wrap=""
        >
          <v-flex-stub
            pr-2=""
            sm3=""
            tag="div"
          >
            <v-select-stub
              item-text="name"
              item-value="uuid"
              items="[object Object]"
              label="State of incorporation"
              outline=""
              single-line=""
              value=""
            />
          </v-flex-stub>
           
          <v-flex-stub
            pr-2=""
            sm3=""
            tag="div"
          >
            <v-select-stub
              item-text="name"
              item-value="id"
              items="[object Object]"
              label="Type"
              outline=""
              single-line=""
              value=""
            />
          </v-flex-stub>
           
          <v-flex-stub
            pr-2=""
            sm3=""
            tag="div"
          >
            <v-select-stub
              item-text="name"
              item-value="uuid"
              items=""
              label="Tags"
              multiple="true"
              outline=""
              single-line=""
              value=""
            />
          </v-flex-stub>
           
          <v-flex-stub
            sm3=""
            tag="div"
          >
            <v-btn-stub
              activeclass="v-btn--active"
              depressed="true"
              large="true"
              tag="button"
              type="button"
            >
              
              Clear all filters
            
            </v-btn-stub>
          </v-flex-stub>
        </v-layout-stub>
      </div>
    </div>
  </div>
   
  <div
    class="px-4"
    style="border-top: 1px solid gainsboro;"
  >
    <v-data-table-stub
      customfilter="function _default(items, search, filter, headers) {
                search = search.toString().toLowerCase();
                if (search.trim() === '') return items;
                var props = headers.map(function (h) {
                    return h.value;
                });
                return items.filter(function (item) {
                    return props.some(function (prop) {
                        return filter(Object(_util_helpers__WEBPACK_IMPORTED_MODULE_7__[\\"getObjectValueByPath\\"])(item, prop, item[prop]), search);
                    });
                });
            }"
      customsort="function _default(items, index, isDescending) {
                if (index === null) return items;
                return items.sort(function (a, b) {
                    var _a, _b;
                    var sortA = Object(_util_helpers__WEBPACK_IMPORTED_MODULE_6__[\\"getObjectValueByPath\\"])(a, index);
                    var sortB = Object(_util_helpers__WEBPACK_IMPORTED_MODULE_6__[\\"getObjectValueByPath\\"])(b, index);
                    if (isDescending) {
                        _a = __read([sortB, sortA], 2), sortA = _a[0], sortB = _a[1];
                    }
                    // Check if both are numbers
                    if (!isNaN(sortA) && !isNaN(sortB)) {
                        return sortA - sortB;
                    }
                    // Check if both cannot be evaluated
                    if (sortA === null && sortB === null) {
                        return 0;
                    }
                    _b = __read([sortA, sortB].map(function (s) {
                        return (s || '').toString().toLocaleLowerCase();
                    }), 2), sortA = _b[0], sortB = _b[1];
                    if (sortA > sortB) return 1;
                    if (sortA < sortB) return -1;
                    return 0;
                });
            }"
      filter="function _default(val, search) {
                return val != null && typeof val !== 'boolean' && val.toString().toLowerCase().indexOf(search) !== -1;
            }"
      headers="[object Object],[object Object],[object Object],[object Object]"
      headertext="text"
      hideactions="true"
      itemkey="uuid"
      items=""
      nexticon="$vuetify.icons.next"
      nodatatext="$vuetify.noDataText"
      noresultstext="$vuetify.dataIterator.noResultsText"
      pagination="[object Object]"
      previcon="$vuetify.icons.prev"
      rowsperpageitems="5,10,25,[object Object]"
      rowsperpagetext="$vuetify.dataTable.rowsPerPageText"
      search=""
      selectall="true"
      sorticon="$vuetify.icons.sort"
      value=""
    />
     
    <div
      class="text-xs-center pt-2"
    >
      <v-pagination-stub
        length="0"
        nexticon="$vuetify.icons.next"
        previcon="$vuetify.icons.prev"
        value="1"
      />
    </div>
  </div>
</div>
`;
